### PR TITLE
MODE-1660: Shell example 

### PR DESCRIPTION
**Pull request by okulikov**
Not for merging yet

### CODE REVIEW

File: modeshape-shell/src/main/java/org/modeshape/example/embedded/ModeShapeExample.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+import java.net.URL;
+import javax.jcr.Node;
+import javax.jcr.Repository;
+import javax.jcr.Session;
+import org.modeshape.common.collection.Problems;
+import org.modeshape.jcr.ModeShapeEngine;
+import org.modeshape.jcr.RepositoryConfiguration;
+
+public class ModeShapeExample {
```
Comment (rhauch) [inline modeshape-shell/src/main/java/org/modeshape/example/embedded/ModeShapeExample.java]: Need a good description of what this examples shows and how it goes about doing that.

File: modeshape-shell/src/main/java/org/modeshape/example/embedded/ModeShapeExample.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            session = repository.login("default");
+
+            // Get the root node ...
+            Node root = session.getRootNode();
+            assert root != null;
+
+            System.out.println("Found the root node in the \"" + session.getWorkspace().getName() + "\" workspace");
+            while(true) {
+                Thread.currentThread().sleep(1000);
+            }
```
Comment (rhauch) [inline modeshape-shell/src/main/java/org/modeshape/example/embedded/ModeShapeExample.java]: This never ends. What is the objective?

**Comment by rhauch** [comment]
@okulikov, what is the objective of the example program? I realize it's incomplete, but do we really want an example and its configuration files in the `modeshape-shell` module?

**Comment by okulikov** [comment]
@rhauch , do you mean `modeshape-shell` module in example git repository? Probably it is not good name for example, should be `modeshape-shell-example`. The purpose is to demonstrate how to incorporate shell and allow to try it. Example builds application with the following directory structure:
- `bin` for OS shell scripts;
-`lib` for modeshape jars;
-`conf` for config files

There are two OS shell scripts: 
- `run.sh` it starts repository and waits for infinite;
-`console.sh` it starts modeshape-shell console;

The sequence is as follows:
- build the applications `mvn clean install`;
- switch to the application root folder `cd target\modeshape-shell`;
-start repository `bin/run.sh`;
-start console `bin/console.sh`;
-execute shell commands (browse, create, modify, remove node, change properties, upload/download binaries, backup/restore, etc).

What would be interested in my opinion for this example is to show how to use the power of OS shell for preparing complex scripts.

**Comment by rhauch** [comment]
@okulikov, I understand that its in the example repository, but I guess I wasn't aware that other applications would embed the shell. Instead, I thought a user would simply run one of the other examples and be able to run the standard shell (which presumably will be an artifact of the build process) to interact with the existing repository.

> What would be interested in my opinion for this example is to show how to use the power of OS shell for preparing complex scripts.

I completely agree. So for this I could see an example in this repository that had one complex (or several minor) scripts that could be applied to a repository.

**Comment by okulikov** [comment]
@rhauch you are actually right with
>  I thought a user would simply run one of the other examples and be able to run the standard shell (which presumably will be an artifact of the build process) to interact with the existing repository.

I just decided to isolate it into dedicated sample. The basis is embedded repository example and everything else just configuration and build process, plus demo shell scripts

**Comment by rhauch** [comment]
@okulikov IMO, we need documentation that describes how a user will use the _actual_ released shell. An example could have a repo pre-configured so they can just start using the shell with it. But I don't understand why this example would be anything but another example repository.

**Comment by okulikov** [comment]
@rhauch , @hchiorean since shell is not completed yet this example allows us to find out in details how released shell will be used by the user. So let's try to identify what is good and what is bad with the shell looking into this sample.

**Comment by hchiorean** [comment]
@okulikov, @rhauch: I also don't understand what the purpose of this example is :smile:  
The console should have 2 parts:

1. the "server" part which facilitates interaction with the repository (which is what https://github.com/ModeShape/modeshape/pull/1538 is for)
2. the "client" part which contains the console shell (command line) and other client-related code which will be a separate project 

(1) should not have any client-facing documentation while (2) should have client-facing documentation but not in the form of code, but `Confluence` style documentation (similar to the the documentation for Wildfly's console for example).

`modeshape-examples` is supposed to be for examples which use the repository by embedding it and IMO the console should never be used like that.

**Comment by okulikov** [comment]
@hchiorean , @rhauch , Let's keep this PR open at least for our development purposes and close it it if we will decide that it shouldn't exist. For now I am using it to play with console features and and for better understanding how it should work.

@hchiorean , @rhauch "client" is a one class project in this case. So what is better:
- create one class project `modeshape-console`;
- merge server and client classes in one project ?

**Comment by hchiorean** [comment]
I vote for having a separate project, even if it will only initially have one class. In addition it will probably have its Maven build, some scripts perhaps ?

**Comment by okulikov** [comment]
I vote for merging. I just can't imagine what kind of other classes except console itself will be here in feature.

**Comment by hchiorean** [comment]
@rhauch ?

**Comment by rhauch** [comment]
@okulikov, I agree with Horia that this is not the appropriate place for experimentation. If the plan is to provide the server-side functionality inside ModeShape, then let's do that and experiment there.

Yes, `modeshape-console` might have just one Java class in the module, but there will certainly be other files (e.g., scripts, starter config file, log file, etc.) that the module has to include in its ZIP artifact that contains client installation. Plus the module's POM is the best place to specify how Maven is to produce the ZIP artifact. Even if the JAR is a runnable JAR (which IMO would be good), I like your suggestion of having scripts that hide the ugliness of the Java executable command.

It is important that the client-side tool contain *only* the JARs and Java classes required for running the client. Even if the client requires some of the same JARs as used on the server (e.g., perhaps `modeshape-common`), they should be utilities and should not affect how the client communicates. Specifically, the client and server should not share any code, such as classes that prescribe the messaging; we should plan for the console to be able to talk to multiple versions of ModeShape.

**Comment by okulikov** [comment]
@rhauch , @hchiorean I see your point.

**Comment by hchiorean** [comment]
@rhauch, @okulikov : so can this be closed and the actual code moved to a separate repo ? (which most likely needs to be created first)

**Comment by rhauch** [comment]
@hchiorean, the client could go in a separate repo or could be just another module with new artifact in the main repo. The server functionality should go in the main repo, tho. Yes?

**Comment by okulikov** [comment]
@hchiorean , closing this PR

**Comment by hchiorean** [comment]
@rhauch: yes

**Pull request closed without merging**